proxy/longpoll_test.go

```diff
@@ -0,0 +1,288 @@
+package proxy
+
+import (
+	stdlibcontext "context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+)
+
+// setting it to a high value, for CI context
+const detectionTimeout = 1200 * time.Millisecond
+
+type (
+	initProxy    func(string) (string, func())
+	testScenario func(*testing.T, initProxy)
+)
+
+type roundTripResponse struct {
+	response *http.Response
+	err      error
+}
+
+// cases:
+// - read request
+// - write header
+// - write response
+// - parallel read/write (probably not supported now)
+
+func backendNotifyingCanceledRequests(
+	readRequest, sendHeader, sendResponse bool,
+	notifyIncoming chan<- <-chan struct{},
+) (url string, closeServer func()) {
+	quit := make(chan struct{})
+	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		canceled := make(chan struct{})
+
+		select {
+		case notifyIncoming <- canceled:
+		case <-quit:
+			return
+		}
+
+		if readRequest {
+			buf := make([]byte, 4)
+			for {
+				_, err := r.Body.Read(buf)
+				if err == io.EOF {
+					break
+				}
+
+				if err != nil {
+					close(canceled)
+					return
+				}
+			}
+		}
+
+		if sendHeader {
+			// in practice, this doesn't block
+			w.WriteHeader(http.StatusOK)
+			w.(interface{ Flush() }).Flush()
+		}
+
+		for loop := true; loop; {
```

Comment:
why do you need `loop` and `for {}` is not enough ?

Reply:
not needed, at some point this loop had a break in a select, but even then, this is not the right way, it's WIP garbage

Reply:
fixed

Reply:
this was not changed